---
title: "Seurat mapping of human L2/3 excitatory Patch-seq cells"
author: "Jeremy Miller"
date: "March 27, 2020"
output: html_notebook
---

This notebook compares human FACs and Patch-seq data.  Prior to data visualization (like what was done in mouse), we also map the Patch-seq cells to the dissociated nuclei reference using Seurat.  These cluster calls are used for downstream analyses.  
  
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
If needed set the working directory:  (e.g., `setwd("C:/Users/jeremym/Desktop/patchseq_human_L23")`).  
  
## Prepare the data  
  
First load the required libraries and options and set directory locations.  
  
```{r load_libraries}
## Load libraries
suppressPackageStartupMessages({
  library(genesorteR)
  library(feather)
  library(dplyr)
  library(gplots)
  library(Seurat)
  library(ggplot2)
  library(cowplot)
  library(scrattch.hicat)
  library(scrattch.vis)
  library(future)
  library(ggbeeswarm)
  library(mfishtools) # getBetaScore
  library(pheatmap)
  library(VENcelltypes)
  library(WGCNA)
  library(matrixStats)
  library(Matrix)
  library(umap)
})
options(stringsAsFactors=FALSE)

## Prepare for extra memory usage
plan("multiprocess", workers = 4)
options(future.globals.maxSize = 4000 * 1024^2)
```
  
Next, load the all the relevant data, which was calculated in a previous script.  
  
```{r load data}
load("input_data_sets.RData")
load("complete_facs_data_sets.RData")
load("umap_coordinates.RData")
```  
  
  
## Classify cell types using Seurat  
  
In this section we are going to classify Patch-seq cells using Seurat.  This does NOT require performing clustering, but rather we can take the clusters originally assigned in the MTG paper and use these to map the Seurat data against.  After some gene filtering, we will take a list of a priori genes expected to be differentially expressed between the 5 cell types of interest as the mapping genes rather than taking the ones based on variability (as is typically done with Seurat).  The goal of this is to de-emphasize the cluster size on the results (e.g., if we subsample cells from each cluster, that should have minimal impact on the mapping).  
  
First, let's do some pre-processing to exclude genes that are enriched in glial clusters or that are exceptionally higher in dissociated nuclei (still called "FACS" for convenience) relative to Patch-seq data.  We also want to exclude sex and mitochondrial from consideration as markers.  
  
```{r select_genes_to_exclude}
# First exclude mitochondrial and sex genes
isExclude <- sort(unique(c(sex_genes,mito_genes)))  
excludeGn <- is.element(rownames(datPatch),isExclude)

# Second find glial genes and exclude these from consideration.  
clMTG   = factor(annoMTG_all$cluster_label)
names(clMTG) <- colnames(datMTG_all)
medians = get_cl_medians(datMTG_all,clMTG)
isUsed  = unique(as.character(clMTG))
isGlia  = isUsed[!(grepl("Exc",isUsed)|grepl("Inh",isUsed))]
isUsed  = setdiff(isUsed,isGlia)

maxUsed = apply(medians[,isUsed],1,max)
maxGlia = apply(medians[,isGlia],1,max)
glialGn = maxGlia>maxUsed     # Exclude genes with maximal median expression in glia

# Third, exclude genes with average expression at least four-fold higher in either FACS or patch-seq
platformGn = abs(rowMeans(datMTG)-rowMeans(datPatch))>=2

# Finally, only keep genes that are expressed in both FACS and Patch-seq (log2(CPM+1)>=1 in 1% of the cells)
expressedGn = (rowSums(datMTG>=1)>(0.01*dim(datMTG)[2]))&(rowSums(datPatch>=1)>(0.01*dim(datPatch)[2]))

keepGenes <- (!(excludeGn|glialGn|platformGn))&(expressedGn)
mean(keepGenes)
```
  
Only these genes will be used in the clustering and visualization analysis below.  
  
  
Data and meta-data set-up.  
  
```{r setup_seurat_objects}
## Basic data and metadata set-up
brain.data     <- cbind(datMTG[keepGenes,],datPatch[keepGenes,])  # Include only genes subsetted above
brain.metadata <- data.frame(set=c(rep("FACs",dim(datMTG)[2]),rep("PatchSeq",dim(datPatch)[2])),
 celltype = c(annoMTG$cluster_label,rep("Patch-seq",dim(annoPatch)[1])), depth = c(annoMTG$scaled_depth,annoPatch$scaled_depth))
rownames(brain.metadata) <- colnames(brain.data)

## Construct data set lists
brain      <- CreateSeuratObject(counts = brain.data, meta.data = brain.metadata)
brain.list <- SplitObject(object = brain, split.by = "set")
```
  
  
Calculate proportions and beta score for variable gene selection.  
  
```{r proportions and beta}
cl        <- setNames(annoMTG$cluster_label,colnames(datMTG))
propExpr  <- get_cl_prop(datMTG[keepGenes,],cl)
betaScore <- getBetaScore(propExpr)
betaOut   <- data.frame(Gene=rownames(datMTG)[keepGenes],BetaScore=betaScore)
betaOut   <- betaOut[order(-betaScore),]
```
  
  
Variable feature selection (beta score replaces the default Seurat method and is also mostly independent of the Patch-seq cells).  
  
```{r variable gene selection}
nGenes      <- 2000
varFeatures <- betaOut$Gene[1:nGenes]

for (i in 1:length(x = brain.list)) {
  VariableFeatures(brain.list[[i]]) <- varFeatures
}
```
  
  
Now let's classify all of the patch-seq data and map it into the MTG space using Seurat 3.0.  **This short section is where the mapping actually happens!**  
  
```{r map patch-seq data}
set.seed(10)  # I don't think this is necessary, but just in case...
dims          <- 30
brainFACs     <- brain.list[["FACs"]]
patch.query1  <- brain.list[["PatchSeq"]]
patch.anchors <- FindTransferAnchors(reference = brainFACs, query = patch.query1, dims = 1:dims, verbose=FALSE, npcs=dims)
predictions   <- TransferData(anchorset = patch.anchors, refdata = brainFACs$celltype, dims = 1:dims, verbose=FALSE)
patch.query1  <- AddMetaData(object = patch.query1, metadata = predictions)
patch.query1$celltype <- patch.query1$predicted.id
```
  
  
How do the proportions compare with FACS?  
  
```{r proportions compare with FACS,fig.height=6,fig.width=6}
facsCount <- table(brain.metadata$celltype[1:dim(datMTG)[2]])
seurCount <- table(factor(patch.query1$celltype,levels=names(facsCount)))

plot(as.numeric(facsCount),as.numeric(seurCount),pch=20,cex=2,xlab="FACS",ylab="Patch-seq")
cor(as.numeric(facsCount),as.numeric(seurCount))^2 
```
  
There is reasonably good agreement in both.  This is consistent with random Patch-seq sampling by cell type.  
  
Next, let's make a combined brain set.  *Note that this is for visualization and for comparison with other features, but that FACs reference alone was used for mapping.*  
  
```{r combined brain set in Seurat}
set.seed(10)  # I don't think this is necessary, but just in case...
## Set up a integrated data set using Patch-seq and FACs data
brain.anchors    <- FindIntegrationAnchors(object.list = brain.list, dims = 1:dims, anchor.features = varFeatures, verbose = FALSE)
  # Note that we are providing anchor features
brain.integrated <- IntegrateData(anchorset = brain.anchors, dims = 1:dims, verbose = FALSE)

## Run the main integration analysis
DefaultAssay(object = brain.integrated) <- "integrated"
brain.integrated <- ScaleData(object = brain.integrated, verbose = FALSE)
brain.integrated <- RunPCA(object = brain.integrated, npcs = dims, verbose = FALSE)
```
  
  
Predict UMAP coordinates of patch-seq data given the UMAP coordinates from FACs.  
  
```{r calculate umap via umap library}
data_patch <- FetchData(brain.integrated, vars=paste0("PC_",1:dims), cells=colnames(datPatch), slot="scale.data") 
data_FACS  <- FetchData(brain.integrated, vars=paste0("PC_",1:dims), cells=colnames(datMTG), slot="scale.data")

# Run umap from umap library
umap_FACs  <- umap(data_FACS) # Predict does not work with method="umap-learn" as far as I can tell
# Replace values with values from Seurat umap in initial analysis
umap_FACs$layout <- as.matrix(umap_human)
# Predict patch-seq locations using "predict" function
umap_patch <- predict(umap_FACs,data_patch)
```
  
  
Calculate UMAP using Seurat.  *These values will be over-written with the UMAP values calculated above.*  
  
```{r calculate umap}
brain.integrated <- RunUMAP(object = brain.integrated, reduction = "pca", dims = 1:dims, verbose = FALSE)

# Update the meta-data with cluster calls from above 
brain.integrated@meta.data$celltypeNew <- brain.integrated@meta.data$celltype
brain.integrated@meta.data$celltypeNew[brain.metadata$set=="PatchSeq"] = predictions$predicted.id

# Update the umap coordinates from above.
brain.integrated@reductions$umap@cell.embeddings <- rbind(umap_FACs$layout,umap_patch)
```
  
  
Plot the results in UMAP space.  
  
```{r plot UMAP for merged analysis, echo=FALSE, fig.height=14, fig.width=14}
xl <- range(umap_human[,1])
yl <- range(umap_human[,2])
colsH <- unique(annoMTG$cluster_color)[match(sort(unique(annoMTG$cluster_label)),unique(annoMTG$cluster_label))]
p1 <- DimPlot(object = brain.integrated, group.by = "set", reduction = "umap", do.return = TRUE, 
              pt.size = 1) + ggtitle("Platform (green=patchseq)") +xlim(xl) + ylim(yl) + NoLegend()
p2 <- DimPlot(object = brain.integrated, group.by = "celltype",  reduction = "umap", 
      do.return = TRUE, pt.size = 1, label=TRUE, label.size = 2,cells=colnames(datMTG), cols=colsH) + NoLegend() + 
      ggtitle("FACS cells") + xlim(xl) + ylim(yl)
p3 <- DimPlot(object = brain.integrated, group.by = "celltypeNew",  reduction = "umap", 
      do.return = TRUE, pt.size = 2, label=FALSE, label.size = 2,cells=colnames(datPatch), cols=colsH) + NoLegend() + 
      ggtitle("Seurat-called types") + xlim(xl) + ylim(yl)
p4 <- DimPlot(object = brain.integrated, group.by = "celltype",  reduction = "umap", 
      do.return = TRUE, pt.size = 2, label=TRUE, label.size = 2,cells=colnames(datPatch)) + NoLegend() + 
      ggtitle("Original Seurat types based on all data") + xlim(xl) + ylim(yl)
plot_grid(p1, p2, p3, p4, ncol=2)
ggsave("UMAP_L23human.pdf",height=11,width=11)
```
  
Plot the PCs for reference.  
  
```{r pcs plot2, fig.height=7,fig.width=12}
features = paste0("PC_",1:6)
FeaturePlot(object = brain.integrated, reduction = "umap", pt.size = 0.5, features = features, ncol=3) + xlim(xl) + ylim(yl)
```
  
  
Calculate the first PC and gene loadings for FREM3 only for use in other Figures.  
  
```{r FREM3 pcs, fig.width=5,fig.height=5}
kpFREM <- predictions$predicted.id=="Exc L2-3 LINC00507 FREM3"
kpCell <- colnames(datPatch)[kpFREM]

nbt    <- CreateSeuratObject(datPatch[keepGenes,kpCell])
nbt    <- FindVariableFeatures(nbt, selection.method = "vst", nfeatures = 1000, mean.cutoff = c(1,Inf), verbose = FALSE)
datTmp <- as.matrix(FetchData(nbt,vars = VariableFeatures(nbt), slot="counts"))
pcFREM <- prcomp(datTmp)

pcTmp2 <- FetchData(brain.integrated,"PC_1",kpCell)
main   <- signif(cor(pcFREM$x[,1],pcTmp2[,1]),3)
plot(pcFREM$x[,1],pcTmp2[,1], xlab="FREM2 PC1", ylab="Seurat Patch-seq PC1", main=paste("Cor =",main))
loading<- sort(pcFREM$rotation[,1])
write.csv(data.frame(gene=names(loading),loading=loading),"frem3_pc1loading.csv",row.names=FALSE)
write.csv(data.frame(cell=kpCell,PC1=pcFREM$x[,1]),"frem3_pcs.csv",row.names=FALSE)
```
  
As a sanity check, we see good (but not perfect) correlation between PC1 from all the data and from FREM3 alone (which includes >50% of the cells and has high heterogeneity).  
  
Plot some qc metrics for the patch-seq cells only.  These plots are mostly sanity check and may not explicitly be included in the manuscript.  
  
```{r qc plot2, fig.height=7,fig.width=12}
isEpilepsy <- annoPatch$medical_conditions=="Epilepsy"
biTmp  <- subset(brain.integrated, cells = colnames(datPatch))
confidence <- predictions$prediction.score.max
biTmp@meta.data <- cbind(biTmp@meta.data,as.data.frame(annoPatch),confidence,isEpilepsy)
features2 <- c("L23_depth_normalized","nFeature_RNA","confidence","marker_sum_norm","isEpilepsy","isTCX")
FeaturePlot(object = biTmp, reduction = "umap", pt.size = 2, features = features2, ncol=3) + xlim(xl) + ylim(yl)
ggsave("qc_feature_umap_L23human.pdf",height=8,width=14)
```
  
Most of the cells appear to be quite high quality.  However, it is interesting to note that the handful of cells on the main island mapping to COL22A1 seem to be on the low end of the NMS score and expressed genes, although not necessarily of the confidence.  
  
## Compare properties with cluster calls and output results  
  
First, we need to merge the meta-data into a single matrix.  
  
```{r merge metadata}
tAnnoPatch <- as.data.frame(annoPatch)
rownames(tAnnoPatch) <- tAnnoPatch$SpecimenID

umap  <- FetchData(object = brain.integrated, vars = paste0('UMAP_',1:2)) 
umap  <- umap[colnames(datPatch),]
colnames(umap) <- gsub("UMAP_","tUMAP_",colnames(umap))

tPcs  <- FetchData(object = brain.integrated, vars = paste0('PC_',1:dims))
tPcs  <- tPcs[colnames(datPatch),]
colnames(tPcs) <- gsub("PC_","tPC_",colnames(tPcs))

metaPatch1 <- data.frame(SpecimenID = rownames(tAnnoPatch), SeuratMapping = patch.query1$predicted.id, 
                         SeuratMappingScore = patch.query1$prediction.score.max, scaled_depth = annoPatch$scaled_depth)
metaPatch  <- cbind(metaPatch1,tAnnoPatch, umap, tAnnoPatch, tPcs)
metaPatch  <- metaPatch[,unique(colnames(metaPatch))]
```
  
  
Ouput the results.  These results are used for later manuscript figures.  
  
```{r output metadata}
write.csv(metaPatch,"IVSCC_excitatory_L23_March2020.csv",row.names=FALSE)
```  
  
  
Now plot the results and significance values.  
  
```{r plot correpondence with everything, warning=FALSE, fig.width=7,fig.height=7}  
kpSampP2 <- 1:dim(metaPatch)[1]  # USE ALL SAMPLES

classes <- NULL
for (i in 1:dim(metaPatch)[2]) classes <- c(classes,class(metaPatch[,i])[1])
annoPatchNum <- metaPatch[kpSampP2,(classes!="character")&(colSums(!is.na(metaPatch))>50)]
sCluster2    <- predictions$predicted.id
sConfidence2 <- predictions$prediction.score.max

getAnovaPvalforApply <- function(x,varLabels,varWeights=NULL){
  anovadat  = as.data.frame(cbind(varLabels,x))
  aov.out   = summary(aov(as.numeric(anovadat[,2])~anovadat[,1],data=anovadat,weights=varWeights))  
  return(aov.out[[1]]$'Pr(>F)'[1])
}
anovaPval    <- apply(t(annoPatchNum),1,getAnovaPvalforApply,sCluster2,sConfidence2)

plots <- list()
N     <- dim(annoPatchNum)[2]
for (i in 1:10){
 j    <- order(anovaPval)[i]
 dat  <- annoPatchNum[,j]
 kp   <- !is.na(dat)
 p    <- ggplot(mapping=aes(x=sCluster2[kp], y=dat[kp], size=sConfidence2[kp])) + 
    geom_boxplot() +
    geom_beeswarm(aes(color=sCluster2[kp]))+ 
    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1)) + 
	theme(legend.position="none") + labs(x='',y="Numeric value") + 
	ggtitle(paste0(names(anovaPval[j]),", ANOVA p=",signif(anovaPval[j],3))) 
 plots[[i]] <- ggplotGrob(p)
}
plots2 <- plots
class(plots2) <- c("arrangelist", class(plots2)) 
plots2

for (i in 11:N){
 j    <- order(anovaPval)[i]
 dat  <- annoPatchNum[,j]
 kp   <- !is.na(dat)
 p    <- ggplot(mapping=aes(x=sCluster2[kp], y=dat[kp], size=sConfidence2[kp])) + 
    geom_boxplot() +
    geom_beeswarm(aes(color=sCluster2[kp]))+ 
    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1)) + 
	theme(legend.position="none") + labs(x='',y="Numeric value") + 
	ggtitle(paste0(names(anovaPval[j]),", ANOVA p=",signif(anovaPval[j],3))) 
 plots[[i]] <- ggplotGrob(p)
}
class(plots) <- c("arrangelist", class(plots))  
ggsave("allInfo_vs_cluster_L23exc_seurat.pdf", plots)
```  
  
There are a very small number of cells that seem to have obviously incorrect mapping based on Seurat, but the vast majority of these results look reasonable.  This is an important point.  While there is some small probability of incorrect mapping of individual cells, the overall cell type-level results should be robust to these handful of potential mis-assignments.  
  
### Differential expression: methods vs. species
  
For this final section, we want to compare the number of differentially-expressed genes by method vs. by species to show that these results are reasonably consistent.  
  
First, define relevant trimmed mean variables.  Since we discuss this before discussing the mapping in the paper, we will use all patch-seq cells.  
  
```{r define medians}
kpGn <- intersect(rownames(datMTG),rownames(datVISp))
meanHomoH <- apply(datMTG,1,median)
meanHomoM <- apply(datVISp,1,median)
meanCarmH <- apply(datMTG[,annoMTG$cluster_label=="Exc L3-4 RORB CARM1P1"],1,median)
meanHomoP <- apply(datPatch,1,median)
plotData  <- data.frame(gene=kpGn, 
                        Human_L23_FACS = meanHomoH[kpGn],
                        Mouse_L23_FACS = meanHomoM[kpGn],
                        Human_CARM1P1_FACS = meanCarmH[kpGn],
                        Human_L23_PatchSeq = meanHomoP[kpGn])
```
  
  
Now plot the correlations.  
  
```{r plot correlations, fig.height=4,fig.width=15}
b <- c(1,5,15,50,150)
bin <- function(x) x + stat_bin_2d(bins=72) + xlim(0,15) + ylim(0,15) +
  scale_fill_gradient(name="count", trans="log2", breaks=b, labels=b, limits=c(1,150)) + 
  geom_abline(intercept=-2, slope=1, linetype="dashed", color="grey", size=1.5) + 
  geom_abline(intercept=2, slope=1, linetype="dashed", color="grey", size=1.5) +
  geom_abline(intercept=0, slope=1, color="black",size=1.5)
p1 <- bin(ggplot(plotData, aes(Mouse_L23_FACS,Human_L23_FACS))) + 
  ggtitle(paste("R =",signif(cor(meanHomoH[kpGn],meanHomoM[kpGn]),2))) + NoLegend()
p2 <- bin(ggplot(plotData, aes(Human_L23_PatchSeq,Human_L23_FACS))) +
  ggtitle(paste("R =",signif(cor(meanHomoH[kpGn],meanHomoP[kpGn]),2))) + NoLegend()
p3 <- bin(ggplot(plotData, aes(Human_CARM1P1_FACS,Human_L23_FACS))) +
  ggtitle(paste("R =",signif(cor(meanHomoH[kpGn],meanCarmH[kpGn]),2))) + NoLegend()
p4 <- bin(ggplot(plotData, aes(Human_CARM1P1_FACS,Human_L23_FACS))) + ggtitle("FOR COLOR BAR")
plot_grid(p1, p2, p3, p4, ncol=4)
ggsave("homology_scatterplots.pdf",height=4,width=15)
```
  
  
Determine how many DEX genes (as defined by log2(FC)>2).  
  
```{r dex genes}
dex <- function (x,y,val=2) print(c(sum(x>=(y+val)),sum(y>=(x+val)),length(x)))
dex(meanHomoH[kpGn],meanHomoM[kpGn])
dex(meanHomoH[kpGn],meanHomoP[kpGn])
dex(meanHomoH[kpGn],meanCarmH[kpGn])
dex(meanHomoH[],meanHomoP[])
dex(meanHomoH[],meanCarmH[])
```
  
  
The majority of DEX genes are homologs to mouse, so we will stick with this list for now.  Now output the lists of DEX genes for input into toppgene.  
  
```{r}
write.csv(plotData,"median_expression_for_GOenrichment.csv",row.names=FALSE)
```
  
  
### View and plot common FACs and Patch-seq markers
  
Data integration only ensures that the cells are matched, but does not say anything about agreement in gene expression.  While the method does adjust gene expression levels, in our experience these adjusted levels are often too divorced from the initial values to be of use.  We next seek to identify common marker genes across data sets using the log2(CPM+1) data as a starting point.  **We provide code below for several different stratgies for doing this for visibility, but we note that the strategy we ended up using for this analysis in the manuscript is the section regarding `genesorteR` towards the end.**  
  
Identify cluster averages and gene specificity across data sets, using the Seurat cluster calls as the subclass definitions.  
  
```{r find subclass based information per data set}
## Use cell type calls from Seurat as the subclass labels
ctFACs  <- subset(brain.integrated,cells = colnames(datMTG))@meta.data$celltypeNew
names(ctFACs) <- colnames(datMTG)
ctPatch <- subset(brain.integrated,cells = colnames(datPatch))@meta.data$celltypeNew
names(ctPatch) <- colnames(datPatch)
clustsH <- c("Exc L2 LAMP5 LTK", "Exc L2-4 LINC00507 GLP2R", "Exc L2-3 LINC00507 FREM3",
             "Exc L3-4 RORB CARM1P1", "Exc L3-5 RORB COL22A1") 

## Calculate trimmed averages across data sets
dats <- list(2^datMTG-1,2^datPatch-1)  # CONVERT BACK TO LINEAR SPACE
cts  <- list(ctFACs,ctPatch)
avg  <- list()
for (i in 1:2){
  avg[[i]] <- do.call("cbind", tapply(names(cts[[i]]), cts[[i]], function(x) apply(dats[[i]][,x],1,mean,trim=0.25))) 
  rownames(avg[[i]]) <- rownames(dats[[i]])
  avg[[i]] <- avg[[i]][,clustsH]
}


```
  
  
Next, find the correlation between data sets and identify genes that are both specific to a small number of subtypes and also correlated across all data sets.  
  
```{r common marker genes, warning=FALSE}
## Calculate correlations across data sets
corMin <- avg[[1]][,1]*0+1
for (i in 2) for(j in 1:(i-1)){
 datTmp <- cbind(avg[[i]],avg[[j]])
 corMin <- pmin(corMin,apply(datTmp,1,function(x) cor(x[1:5],x[6:10])))
}
corMin[is.na(corMin)]=-1
```
  
  
As a first step we need to calculate some statistics on these clusters, including means and proportions.  
  
```{r calculate statistics}
## FACS statistics
exprThresh  <- 1
propFACS    <- do.call("cbind", tapply(names(ctFACs), ctFACs, function(x) rowMeans(datMTG[,x]>exprThresh))) 
meanFACS    <- avg[[1]] # Calculated above
rownames(propFACS) <- rownames(meanFACS) <- rownames(datMTG)  
propFACS    <- propFACS[,clustsH]
meanFACS    <- meanFACS[,clustsH]

## Patch statistics
propPatch   <- do.call("cbind", tapply(names(ctPatch), ctPatch, function(x) rowMeans(datPatch[,x]>exprThresh))) 
meanPatch   <- avg[[2]] # Calculated above 
rownames(propPatch) <- rownames(meanPatch) <- rownames(datPatch)  
propPatch   <- propPatch[,clustsH]
meanPatch   <- meanPatch[,clustsH]
```
  
  
Next, find the top marker genes for the ET cluster in each data set (Exc FEZF2 GABRQ in FI and Exc L4-5 FEZF2 SCN4B in MTG).  Note that we are using slightly different statistical criteria and comparison types as we did in the cross-species analysis, and therefore the resulting gene lists do not perfectly align.  
  
```{r identify top ET genes}
# FACS data
wmeF <- apply(meanFACS,1,which.max)
wmpF <- apply(propFACS,1,which.max)
fceF <- apply(meanFACS,1,function(x) diff(range(-sort(-x))))
pceF <- apply(propFACS,1,function(x) diff(range(-sort(-x))))

# Patch data
wmeP <- apply(meanPatch,1,which.max)
wmpP <- apply(propPatch,1,which.max)
fceP <- apply(meanPatch,1,function(x) diff(range(-sort(-x))))
pceP <- apply(propPatch,1,function(x) diff(range(-sort(-x))))

# Rough markers
kp <- (wmeP==wmeF)&(wmpP==wmeF)&(wmeP==wmpF)&(corMin>0.75)&(fceF>=3)&(fceP>=3)&(pceF>0.6)&(pceP>0.6)
gn <- names(sort(-corMin[kp]))

# Final markers
mark <- NULL
for (i in 1:5){
  tmp  <- gn[wmpF[gn]==i]
  mark <- c(mark,tmp[1:min(5,length(tmp))])
}
mark <- unique(c(mark,"LTK", "GLP2R", "FREM3", "CARM1P1", "COL22A1"))
```  
  
  
Plot top markers.  
  
```{r plot to screen common marker genes, fig.height=8,fig.width=8}
##  Order the genes and plot the results
plotGn <- mark
tmp <- meanFACS[plotGn,]+meanPatch[plotGn,] 
tmp <- tmp/rowSums(tmp)
ord <- order(-apply(tmp,1,which.max)*10,rowMeans(t(apply(tmp,1,cumsum))))
mains <- c("FACs","Patch-seq")
mark  <- plotGn[ord]

for (i in 1:2){
 plotVal <- avg[[i]][mark,]
 plotVal <- plotVal/apply(plotVal,1,max)
 heatmap(plotVal,Rowv=NA,Colv=NA, main=mains[i],scale="none");
}
```
  
  
Now plot these as a violin plot, which is a fairer representation of the data.  
  
```{r Data organization}
data    <- cbind(datMTG,datPatch)
data.df <- cbind(sample_name = colnames(data), as.data.frame(2^t(data[mark,])-1))
spl     <- function(x) as.character(sapply(strsplit(x," "),function(x) x[4]))
clF     <- setNames(paste(1:5,"FACS",spl(clustsH)),clustsH)
clP     <- setNames(paste(1:5,"Patch",spl(clustsH)),clustsH)
cn      <- c("cluster_label","cluster_color")
clTmp   <- c(clF[clustsH],clP[clustsH])
cols    <- makeColorsUnique(setNames(rep(annoMTG$cluster_color[match(clustsH,annoMTG$cluster_label)],2),clTmp))
anno    <- data.frame(sample_name=colnames(data),cluster_label=c(clF[ctFACs],clP[ctPatch]))
anno$cluster_color <- cols[anno$cluster_label]
anno$cluster_id    <- as.numeric(as.factor(anno$cluster_label))
```
  
  
```{r plot top genes in MTG, fig.width=4,fig.height=9}
group_violin_plot(data.df, anno, grouping ="cluster", genes=mark[length(mark):1], log_scale = TRUE, 
                       max_width = 20, label_height = 15)
ggsave("violinPlots_markerGenes.pdf", width = 4, height = 9)
```
  
These results look reasonable but not visually impressive, so we are going to omit this result from the manuscript for now.  
  
Use genesortR to identify genes differentially expressed between FREM3, COL22A1, and CARM1P1 for DEX analysis.  These gene lists are compared against gene lists calculated the same way from the FACS data to determine consisency between methods.  This comparison was done from these gene lists output and not in R.  
  
```{r}
clp = setNames(metaPatch$SeuratMapping,colnames(datPatch))

#filter cells
cl3       <- c("Exc L2-3 LINC00507 FREM3", "Exc L3-4 RORB CARM1P1", "Exc L3-5 RORB COL22A1")
kp3       <- is.element(clp,cl3)&(annoPatch$scaled_depth>=0.575)
kp3[is.na(kp3)] = FALSE
propExpr3 <- get_cl_prop(datPatch[,kp3],clp[kp3])


#very roughly filter genes
betaScore3 <- getBetaScore(propExpr3[rowSums(propExpr3[,cl3])>0.3,cl3])
kpGenes3   <- names(sort(-betaScore3))[1:10000]
write(sort(kpGenes3[1:300]),"human_markers_beta300p.txt")

#get specificity scores for each cell type
sg = sortGenes(datPatch[kpGenes3,kp3], clp[kp3])

head(sg$specScore) #specificity scores for each gene in each cluster

#define a small set of markers
mm = getMarkers(sg, quant = 0.7)

#cluster genes and make a heatmap
pp = plotMarkerHeat(sg$inputMat, sg$inputClass, mm$markers, clusterGenes=TRUE, outs = TRUE)
pdf("dex_heatmap_genesorteRp.pdf",height = 9,width=4)
pp$pheat
dev.off()

pp = plotMarkerHeat(sg$inputMat, sg$inputClass, kpGenes3[1:300], clusterGenes=TRUE, outs = TRUE)
pdf("dex_heatmap_beta300p.pdf",height = 9,width=4)
pp$pheat
dev.off()

write(sort(mm$markers),"human_markers_genesorteRp.txt")
```
  
The genes output by `genesorteR` look like better markers for this context than the top beta genes, so that is the set that we will use for the manuscript.  
  
Output session information.  
  
```{r sessionInfo}
sessionInfo()
```
  
  